Сортировка массива строк

vharlamov opened the pull request:
Очень трудно было догадаться, как прописать смешанную локаль. Нигде этого нет, догадался методом тыка, по наитию.

vharlamov committed:
Сортировка массива строк

Очень трудно было догадаться, как прописать смешанную локаль. Нигде этого нет, догадался методом тыка, по наитию.

vharlamov committed:
Merge branch 'master' of https://github.com/js-tasks-ru/js-20201211_valera545

vharlamov committed:
Pick

npm test ругался на второй цикл, которым я добавлял недостающие поля. Непонятно, но убрал его (закомментил).

vharlamov committed:
Merge branch 'master' of https://github.com/js-tasks-ru/js-20201211_valera545

vharlamov committed:
Omit

jsru commented:
Поздравляю, все задачи решены абсолютно верно
```diff
sortStrings
+ should return sorted by "asc" array of strings
+ should return a new sorted array
+ should return sorted by "desc" array of strings in "en" locale
+ should correctly sort language-specific characters in "ru" locale
+ should correctly sort strings started from uppercase
+ should correctly sort strings for mixed "en" and "ru" locales
```
```diff
pick
+ should return new object with one passed field
+ should return new object with a few passed field
+ should return an empty object if passed fields does not found
```
```diff
omit
+ should return a new object without passed field
+ should return a new object without few passed fields
+ should return initial object if passed fields doesn't found
```

Добавляю преподавателя (@dosandk) для код-ревью, посмотрим что он скажет :)

jsru commented:
Решение было обновлено, посмотрим что скажет @dosandk

jsru commented:
Решение было обновлено, посмотрим что скажет @dosandk

dosandk commented on 02-javascript-data-types/1-sort-strings/index.js:
```diff
@@ -5,5 +5,11 @@
-
+    let tmpArr = arr.slice();
+    if( param == 'asc' ) {
+        return tmpArr.sort( (a, b) => a.localeCompare( b, 'ru-en', { caseFirst:'upper', sensitivity:'case', usage:'sort' } ) );
```
строка 10 дублирует строку 12 за исключением порядка аргументов в методе `localeCompare` - попробуйте избавиться от дублирования.

проверьте использование `let` - возможно при объявлении переменной в данном решении нужен `const`?

dosandk commented on 02-javascript-data-types/1-sort-strings/index.js:
```diff
@@ -5,5 +5,11 @@
+        return tmpArr.sort( (a, b) => a.localeCompare( b, 'ru-en', { caseFirst:'upper', sensitivity:'case', usage:'sort' } ) );
+    } else  {
+        return tmpArr.sort( (a, b) => a.localeCompare( b, 'ru-en', { caseFirst:'lower', sensitivity:'case', usage:'sort' } ) ).reverse();
+    };
```
Подумайте как исправить решение таким образом чтобы избавиться от `reverse` и сортировать данные сразу в верном порядке

dosandk commented on 02-javascript-data-types/2-pick/index.js:
```diff
@@ -5,5 +5,18 @@
+        newList[ key ] = value;
+      }
+    }
+    // for( let field of fields ) {
```
никогда, никогда не оставляйте закомментированный код - это очень плохая практика.
подробнее расскажу на занятии

dosandk commented on 02-javascript-data-types/2-pick/index.js:
```diff
@@ -5,5 +5,18 @@
-};
+    let newList = {};
+  
+    for( let [key, value] of Object.entries( obj ) ) {
```
good 👍

dosandk commented on 02-javascript-data-types/2-pick/index.js:
```diff
@@ -5,5 +5,18 @@
 export const pick = (obj, ...fields) => {
-
-};
+    let newList = {};
```
может `const`? как считаете?

dosandk reviewed (changes requested):
необходимо исправить решение `sortStrings`. Проверить использование `let` и `const`

vharlamov committed:
sortString

Исправлено:
1. const вместо let;
2. убран reverce;

Не исправлено:
1. Не вижу возможности избавиться от дублирования строки, сохранив при этом правильнуюю расстановку заглавных и строчных. Помимо порядка аргументов в localeCompare изменён параметр caseFirst с 'upper' на 'lower', который обеспечивает эту расстановку. Если это неважно, тогда исправлю.

jsru commented:
Решение было обновлено, посмотрим что скажет @dosandk

vharlamov committed:
sortString

Исправлено: устранено дублирование строки в условиях.

jsru commented:
Решение было обновлено, посмотрим что скажет @dosandk

vharlamov committed:
sortString

Ещё чуть поправил — объявил все переменные сразу в одной строке.

jsru commented:
Решение было обновлено, посмотрим что скажет @dosandk

vharlamov committed:
sortString

Заменил const tmpArr = arr.slice() на const tmpArr = [...arr].
Нигде такого до сих пор не встречал ))

jsru commented:
Решение было обновлено, посмотрим что скажет @dosandk
